### PR TITLE
adding retry logic for retriable errors

### PR DESCRIPTION
**Pull request by rcohenma**

### CODE REVIEW

File: crud/helpers.go [hunk 1]
```diff
@@ -136,14 +138,26 @@ func LoadBalancerResourceGet(s BaseCrud, workReq *baremetal.WorkRequest) (id str
...
 }
 
+func sleepHelper(retryNum uint) {
+	secondsToSleep := 1 << retryNum
+	log.Printf("[DEBUG] Got a retriable error. Waiting %d seconds and trying again...", secondsToSleep)
+	time.Sleep(time.Duration(secondsToSleep) * time.Second)
+}
+
+func isRetriableError(error string) bool {
+	match, _ := regexp.MatchString("status\\s*:\\s*40[134]", strings.ToLower(error))
```
Comment (jodoglevy) [inline crud/helpers.go]: We consider a 400 a retryable error, but not a 404? Are we sure none of the cases being hit where retries were necessary returned 404s?

Comment (josephholsten) [inline crud/helpers.go]: This function needs doc, so I can figure out what the regexp means at 3am ;-)

File: crud/helpers.go [hunk 2]
```diff
@@ -203,9 +235,17 @@ func UpdateResource(d *schema.ResourceData, sync ResourceUpdater) (e error) {
 // () -> Pending -> Deleted.
 // Finally, sets the ResourceData state to empty.
 func DeleteResource(d *schema.ResourceData, sync ResourceDeleter) (e error) {
+	return retriableDeleteResource(d, sync, 1)
+}
+
+func retriableDeleteResource(d *schema.ResourceData, sync ResourceDeleter, retryNum uint) (e error) {
```
Comment (josephholsten) [inline crud/helpers.go]: can we call this `deleteResourceWithRetry`?

File: crud/helpers.go [hunk 3]
```diff
@@ -188,8 +212,16 @@ func ReadResource(sync ResourceReader) (e error) {
 }
 
 func UpdateResource(d *schema.ResourceData, sync ResourceUpdater) (e error) {
+	return retriableUpdateResource(d, sync, 1)
+}
+
+func retriableUpdateResource(d *schema.ResourceData, sync ResourceUpdater, retryNum uint) (e error) {
```
Comment (josephholsten) [inline crud/helpers.go]: can we call this `updateResourceWithRetry`?

File: crud/helpers.go [hunk 4]
```diff
@@ -136,14 +138,26 @@ func LoadBalancerResourceGet(s BaseCrud, workReq *baremetal.WorkRequest) (id str
 	return id, false, nil
 }
 
+func sleepHelper(retryNum uint) {
```
Comment (josephholsten) [inline crud/helpers.go]: how about calling this `exponentialBackoffSleep`

File: crud/helpers.go [hunk 5]
```diff
@@ -136,14 +138,27 @@ func LoadBalancerResourceGet(s BaseCrud, workReq *baremetal.WorkRequest) (id str
...
+
+func createResourceWithRetry(d *schema.ResourceData, sync ResourceCreator, retryNum uint) (e error) {
 	if e = sync.Create(); e != nil {
-		// Check for conflicts and retry
-		// This happens with concurrent volume attachments, etc
-		if strings.Contains(strings.ToLower(e.Error()), "try again later") {
-			log.Println("[DEBUG] Resource creation conflicts with other resources. Waiting 10 seconds and trying again...")
-			time.Sleep(10 * time.Second)
-			e = CreateResource(d, sync)
+		if isRetriableError(e.Error()) && retryNum <= MaxRetries {
```
Comment (codycushing) [inline crud/helpers.go]: Why not pass the error itself and test the code field directly rather than stringifying and regexing on the string.

**Pull request merged**